proto/internal/core/v1/core.proto

```diff
@@ -12,6 +13,28 @@ message RelationTuple {
 
   /** subject is the subject for the tuple */
   ObjectAndRelation subject = 2 [ (validate.rules).message.required = true ];
+
+  /** referencing a caveat that must be enforced over the tuple **/
+  ContextualizedCaveat caveat = 3 [ (validate.rules).message.required = false ];
+}
+
+message ContextualizedCaveat {
+  /** caveat_id is the reference to the internal id of a stored caveat **/
+  uint64 caveat_id = 1 [(validate.rules).uint64.gt = 0];
+
+  /** predefined are arguments used as input during caveat evaluation with a predefined value **/
+  google.protobuf.Struct context = 2 [ (validate.rules).message.required = false ];
+}
+
+message Caveat {
+  /** name represents the globally-unique identifier of the caveat **/
+  string name = 1 [ (validate.rules).string = {
+    pattern : "^(([a-zA-Z0-9_][a-zA-Z0-9/_|-]{0,127})|\\*)$",
+    max_bytes : 128,
+  } ];
+
+  /** expression is the byte representation of a caveat's logic */
```

Comment:
I wonder if we should just make this `DecodedCaveat` and move it in here...

Reply:
honestly it's not clear to me what is the design rationale behind the `impl` proto package. It feels like both would overlap in responsibility.